api/admin.go

```diff
@@ -334,9 +346,70 @@ func (a *API) adminUserDelete(w http.ResponseWriter, r *http.Request) error {
 			return internalServerError("Error recording audit log entry").WithInternalError(terr)
 		}
 
-		if terr := tx.Destroy(user); terr != nil {
-			return internalServerError("Database error deleting user").WithInternalError(terr)
+		if params.IsSoftDelete == "true" {
+			if user.DeletedAt != nil {
+				// user has been soft deleted already
+				return nil
+			}
+			softDeleteId, terr := crypto.GenerateNanoId(5)
+			if terr != nil {
+				return terr
+			}
+			user.Email = storage.NullString(fmt.Sprintf("%s-%x", softDeleteId, sha256.Sum256([]byte(user.GetEmail()))))
+			user.Phone = storage.NullString(fmt.Sprintf("%s-%x", softDeleteId, sha256.Sum256([]byte(user.GetPhone()))))
+			user.EmailChange = fmt.Sprintf("%s-%x", softDeleteId, sha256.Sum256([]byte(user.EmailChange)))
+			user.PhoneChange = fmt.Sprintf("%s-%x", softDeleteId, sha256.Sum256([]byte(user.PhoneChange)))
+			now := time.Now()
+			user.DeletedAt = &now
+			userMetaDataUpdates := map[string]interface{}{}
+			for k, v := range user.UserMetaData {
+				switch t := v.(type) {
+				case string:
+					userMetaDataUpdates[k] = fmt.Sprintf("%s-%x", softDeleteId, sha256.Sum256([]byte(t)))
+				case []byte:
+					userMetaDataUpdates[k] = fmt.Sprintf("%s-%x", softDeleteId, sha256.Sum256(t))
+				}
+			}
+			if terr := tx.UpdateOnly(user, "email", "phone", "email_change", "phone_change", "deleted_at"); terr != nil {
+				return internalServerError("Error soft deleting user").WithInternalError(terr)
+			}
+			if terr := user.UpdateUserMetaData(tx, userMetaDataUpdates); terr != nil {
+				return internalServerError("Error soft deleting user meta data").WithInternalError(terr)
+			}
+
+			identities, terr := models.FindIdentitiesByUser(tx, user)
+			if terr != nil {
+				return internalServerError("Error retrieving identities").WithInternalError(terr)
+			}
+
+			for _, identity := range identities {
+				identity.ProviderId = fmt.Sprintf("%s-%x", softDeleteId, sha256.Sum256([]byte(identity.ProviderId)))
+				if terr := tx.UpdateOnly(identity, "provider_id"); terr != nil {
+					return internalServerError("Error soft deleting identity id").WithInternalError(terr)
+				}
+				identityDataUpdates := map[string]interface{}{}
+				for k, v := range identity.IdentityData {
+					switch t := v.(type) {
+					case string:
+						identityDataUpdates[k] = fmt.Sprintf("%s-%x", softDeleteId, sha256.Sum256([]byte(t)))
+					case []byte:
+						identityDataUpdates[k] = fmt.Sprintf("%s-%x", softDeleteId, sha256.Sum256(t))
+					}
+				}
+				if terr := identity.UpdateIdentityData(tx, identityDataUpdates); terr != nil {
+					return internalServerError("Error soft deleting identity data").WithInternalError(terr)
+				}
```

Comment:
Given the recent PRs, you would also likely need to set the `identity_data` column to `NULL` on all identities belonging to that user.